File: Linked Lists/Inserting a Node Into a Sorted Doubly Linked List.js
'use strict';

const fs = require('fs');

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.replace(/\s*$/, '')
        .split('\n')
        .map(str => str.replace(/\s*$/, ''));

    main();
});

function readLine() {
    return inputString[currentLine++];
}

const DoublyLinkedListNode = class {
    constructor(nodeData) {
        this.data = nodeData;
        this.next = null;
        this.prev = null;
    }
};

const DoublyLinkedList = class {
    constructor() {
        this.head = null;
        this.tail = null;
    }

    insertNode(nodeData) {
        let node = new DoublyLinkedListNode(nodeData);

        if (this.head == null) {
            this.head = node;
        } else {
            this.tail.next = node;
            node.prev = this.tail;
        }

        this.tail = node;
    }
};

function printDoublyLinkedList(node, sep, ws) {
    while (node != null) {
        ws.write(String(node.data));

        node = node.next;

        if (node != null) {
            ws.write(sep);
        }
    }
}

// Complete the sortedInsert function below.

/*
 * For your reference:
 *
 * DoublyLinkedListNode {
 *     int data;
 *     DoublyLinkedListNode next;
 *     DoublyLinkedListNode prev;
 * }
 *
 */
function sortedInsert(head, data) {
    let newNode = new DoublyLinkedListNode(data)
    if (head.data > newNode.data) {
        newNode.next = head
        newNode.next.prev = newNode
        head = newNode
        return head
    } else {
        let curr = head
        let previous = null
        while (curr) {
            if (curr.data >= newNode.data) {
                let temp = curr
                curr = newNode
                previous.next = curr
                curr.prev = previous
                curr.next = temp
                temp.prev = curr
                break;
            } else if (!curr.next && curr.data < newNode.data) {
                curr.next = newNode
                newNode.prev = curr
                break;
            }
            previous = curr
            curr = curr.next
        }
        return head
    }
}

function main() {

File: Linked Lists/Inserting a Node Into a Sorted Doubly Linked List.py
#!/bin/python3

import math
import os
import random
import re
import sys

class DoublyLinkedListNode:
    def __init__(self, node_data):
        self.data = node_data
        self.next = None
        self.prev = None

class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    def insert_node(self, node_data):
        node = DoublyLinkedListNode(node_data)

        if not self.head:
            self.head = node
        else:
            self.tail.next = node
            node.prev = self.tail


        self.tail = node

def print_doubly_linked_list(node, sep, fptr):
    while node:
        fptr.write(str(node.data))

        node = node.next

        if node:
            fptr.write(sep)

# Complete the sortedInsert function below.

#
# For your reference:
#
# DoublyLinkedListNode:
#     int data
#     DoublyLinkedListNode next
#     DoublyLinkedListNode prev
#
#
def sortedInsert(head, data):
    newNode = DoublyLinkedListNode(data)
    if head.data > newNode.data:
        newNode.next = head
        newNode.next.prev = newNode
        head = newNode
        return head
    else:
        current = head
        previous = None
        while current:
            if current.data >= newNode.data:
                temp = current
                current = newNode
                previous.next = current
                current.prev = previous
                current.next = temp
                temp.prev = current
                break;
            elif current.next == None and current.data < newNode.data:
                current.next = newNode
                newNode.prev = current
                break;
            previous = current
            current = current.next 
        return head
    
if __name__ == '__main__':